src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/TempPE/TempPEFileEditHintReceiver.cs

```diff
@@ -0,0 +1,51 @@
+﻿using System;
+using System.Collections.Immutable;
+using System.ComponentModel.Composition;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.VS.Automation;
+using VSLangProj;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.TempPE
+{
+    /// <summary>
+    /// This class handles being notified whenever a source file is edited, in our case by a Single File Generator
+    /// </summary>
+    [Export(typeof(IProjectChangeHintReceiver))]
+    [ProjectChangeHintKind(ProjectChangeFileSystemEntityHint.EditedFileAsString)]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasic)]
+    internal class TempPEFileEditHintReceiver : IProjectChangeHintReceiver
+    {
+        private readonly VSBuildManager _buildManager;
+        private readonly Lazy<ITempPEBuildManager> _tempPEBuildManager;
+
+        [ImportingConstructor]
+        public TempPEFileEditHintReceiver(
+            [Import(typeof(BuildManager))]VSBuildManager buildManager,
+            Lazy<ITempPEBuildManager> tempPEBuildManager)
+        {
+            _buildManager = buildManager;
+            _tempPEBuildManager = tempPEBuildManager;
+        }
+
+        public async Task HintedAsync(IImmutableDictionary<Guid, IImmutableSet<IProjectChangeHint>> hints)
+        {
+            // We don't want to be the one to realise the Lazy<ITempPEManager> so ignore changes in projects that don't need TempPE
+            if (!_buildManager.ProjectNeedsTempPE) return;
```

Comment:
I think here you need to separate the concept from; _ITempPEManager has been created_ vs. _it is actually doing work_. Once you start having a large set of Lazy<>'s, the overhead adds up. In total, the 3 `Lazy<ITempPEBuildManager>` (and MEF baggage) are actually bigger in size than ITempPEBuildManager itself. I would remove this "ProjectNeedsTempPE" concept.

Reply:
I think I just went a bit far here with responding to your comment about not being loaded unless its needed. I can remove this concept easily, it was a late addition anyway, it will just mean there will be a TempPEBuildManager created for every unconfigured project, whether its needed or not. The data flow structures still won't be initialised and collecting data until TempPE is needed though.